### PR TITLE
Re-factor of MOM_ANN

### PR DESCRIPTION
**Pull request by adcroft**
- Moved MOM_ANN.F90 to src/framework/
- Removed unused modules
- Removed unused MOM_memory.h
- Added input and output means which default to 0 and do not need to be present in the weights file
- Gave defaults to means, norms, tests so that they do no need to be present in file
- Added missing array notation "(:)"
- Minor formatting
- Added ANN_allocate, set_layer, set_input_normalization, and set_output_normalization methods to allow reconfiguration during unit tests
- Added ANN_unit_tests with some simple constructed-by-code networks with known solutions
- Added config_src/drivers/unit_tests/test_MOM_ANN.F90 to drive unit tests
- Added  config_src/drivers/timing_tests/time_MOM_ANN.F90 as rudimentary for timing inference
- Added module dox
- Renamed _v1, _v2 etc to labels
- Added interface, ANN_apply, to span ANN_apply_vecvtor_oi and ANN_apply_array_sio
- Alternate versions of code temporarily existed for purpose of evaluating performance

### CODE REVIEW

**Comment by adcroft** [comment]
@Pperezhogin  There's a suggestion we stop using "ANN" to describe MLPs or FFNNs. Perhaps we should change the names in the code now before it becomes part of the main branch?

**Comment by Pperezhogin** [comment]
@adcroft 
By using "ANN", we are consistent with two our recent papers (Sane et al 2023, Perezhogin et al 2025) and early papers on subgrid modeling ([Xie et al 2020](https://journals.aps.org/prfluids/abstract/10.1103/PhysRevFluids.5.054606), [Maulik et al 2019](https://www.cambridge.org/core/journals/journal-of-fluid-mechanics/article/subgrid-modelling-for-twodimensional-turbulence-using-neural-networks/10EDED1AEAA52C35F3E3A3BB6DC218C1), [Beck et al 2019](https://www.sciencedirect.com/science/article/pii/S0021999119306151?casa_token=lqyaEQw3u5EAAAAA:1uJW4KNdMGVuJrHLISKScI9k8k4juivCrMRU4I10Dz2aR89NuZsGyFhtFLxutI3lVspy2ng)), but not consistent with Dhruv et al 2025 (who uses "MLP"). 

By using "ANN", it is unlikely we can mislead anybody. It is reasonable to assume that the most general term, such as ANN, would imply the most common architecture, which is MLP. In terms of applied science, I believe the major problem of ML is an amount of newly emerging different terms while actual neural net architectures are not that diverse. For example, I would be happy if people refer to "FNO" as "ANN in Fourier space". That way a lot of confusion could be eliminated. So, if we stick to only one term ("ANN") instead of producing a multitude of them to refer to the same thing, it is good for us. An opinion on this topic from the viewpoint of purely ML community may be different, but I assume we are not part of that community.

In terms of software, using "MLP" may be reasonable in a case if we want to implement 10+ architectures in the future, and we want clearly distinguish between them. However, even in this case, "ANN" likely cannot be misunderstood as anything different from MLP.

One can get some inspiration in naming convention from pytorch. This package does not have either of these terms: "ANN", "CNN", "MLP", "FFNN", "FNO". This suggests that probably actual neural net architecture and its software implementation are different things. Let's discuss in person.

File: src/framework/MOM_ANN.F90 [hunk 1]
```diff
@@ -163,26 +183,25 @@ end subroutine ANN_end
...
-  type(ANN_CS), intent(in) :: CS !< ANN control structure
-
+  type(ANN_CS), intent(in)      :: CS !< ANN control structure
   real, dimension(CS%layer_sizes(1)), &
-                  intent(in)  :: x !< input [arbitrary]
+                  intent(in)    :: x !< input [arbitrary]
   real, dimension(CS%layer_sizes(CS%num_layers)), &
-                  intent(out) :: y !< output [arbitrary]
-
+                  intent(inout) :: y !< output [arbitrary]
```
Comment (Pperezhogin) [inline src/framework/MOM_ANN.F90]: Just curious, is this change (out to inout) can be explained?

File: src/framework/MOM_ANN.F90 [hunk 2]
```diff
@@ -12,7 +12,17 @@ module MOM_ANN
 !#include <MOM_memory.h>
 
 public ANN_init, ANN_allocate, ANN_apply, ANN_end, ANN_unit_tests
+public ANN_apply_vector_v1, ANN_apply_vector_v2, ANN_apply_vector_v3
+public ANN_apply_vector_v4
+public ANN_apply_array_v1, ANN_apply_array_v2, ANN_apply_array_v3
 public set_layer, set_input_normalization, set_output_normalization
+public ANN_random, randomize_layer
+
+!> Applies linear layer to input data x and stores the result in y with
```
Comment (Pperezhogin) [inline src/framework/MOM_ANN.F90]: Multiple linear layers

File: src/framework/MOM_ANN.F90 [hunk 3]
```diff
@@ -26,6 +36,8 @@ module MOM_ANN
   real, allocatable :: A(:,:) !< Matrix in column-major order
                               !! of size A(output_width, input_width) [nondim]
   real, allocatable :: b(:)   !< bias vector of size output_width [nondim]
+  real, allocatable :: Atranspose(:,:) !< Matrix in column-major order
+                                       !! of size A(output_width, input_width) [nondim]
```
Comment (Pperezhogin) [inline src/framework/MOM_ANN.F90]: Atranspose does not seem to be used in the latest commit

File: config_src/drivers/timing_tests/time_MOM_ANN.F90 [hunk 4]
```diff
@@ -217,15 +217,15 @@ subroutine time_ANN(nlayers, nin, width, nout, nsamp, nits, nxy, impl, label)
   tstd = tstd / real(nsamp) ! convert to mean of squares
   tstd = tstd - tmean**2  ! convert to variance
   tstd = sqrt( tstd * real(nsamp) / real(nsamp-1) ) ! convert to standard deviation
-  flops = ANN%parameters / tmean
+  words_per_sec = ANN%parameters / ( tmean * 1024 * 1024 )
```
Comment (Pperezhogin) [inline config_src/drivers/timing_tests/time_MOM_ANN.F90:177]: Seems that this metric changed its meaning. Before it was measuring number of operators per second (flops) and now it is the memory throughput. I am not sure which metric would be the most relevant as it is unclear a priori if this code will be compute-bounded or memory-bounded. Is this words_per_sec metric supposed to be compared to L1/L2/L3 cache throughput? 

I somewhat find Gflops to be more relevant metric as its range on single CPU core is clearly defined from approximately 0.5Gflops for scalar operators (typically, in scalar code per one floating point operation there are ~5 service operations, [matmul benchmark](https://github.com/Pperezhogin/How-instruction-set-influences-GFLOPS-in-matvec/blob/main/20-core%20Intel%20Xeon%20(Cascade%20Lake).txt)) up to approximately 50-100Gflops for FMA instructions in longest vector registers assuming no memory transfer between registers and cache.  Typical ocean model has 3Gflops performance on average which is much better than scalar code but still far away from the compute bound. I would say a metric of success for ANN module is to be more efficient than ocean model on average, i.e. be in a range of
3Gflops-100Gflops

May be, if we want Gflops, we may need to estimate the number of floating point operations, which is for matmul not the number of parameters in matrix, but approximately twice (one add and one multiply per element).

File: src/framework/MOM_ANN.F90 [hunk 5]
```diff
@@ -0,0 +1,739 @@
...
+  integer :: i
+
+  deallocate(CS%layer_sizes)
+  deallocate(CS%input_means)
+  deallocate(CS%input_norms)
+  deallocate(CS%output_means)
+  deallocate(CS%output_norms)
+
+  do i = 1, CS%num_layers-1
+    deallocate(CS%layers(i)%A)
```
Comment (Pperezhogin) [inline src/framework/MOM_ANN.F90:233]: There is no deallocation of Atranspose and so likely memory leak

File: config_src/drivers/timing_tests/time_MOM_ANN.F90 [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+  tmax = 0.
+  tmean = 0.
+  tstd = 0.
+  asamp = nits ! Most cases below use this
+  aits = nits / nxy ! Most cases below use this
+
+  do samp = 1, nsamp
+    select case (impl)
+      case (0)
+        aits = nits
```
Comment (Pperezhogin) [inline config_src/drivers/timing_tests/time_MOM_ANN.F90:134]: I would remove aits=nits here

**Comment by Pperezhogin** [comment]
Hi @adcroft I am happy to accept this code with a few comments above. 

For future (not this PR). I tried to measure timing (on Perlmutter, AMDs) for ANN with typical use case: 
```
nlayers = 3; nin = 27; layer_width = 20; nout = 3
nxy = 400 (somewhat reasonable size for global ocean model when each MPI rank holds around 20x20 grid points)
```
In this case we have:
```
ANN_apply_vector_orig: "mean": 1.8055E-07
ANN_apply_array_sio: "mean": 9.5918E-08
```
and so the acceleration of new algorithm is about twice (note, it may depend on the cluster). I would say, the inference time per grid point is very close to pytorch that was `1E-7 s` per grid point on 1 CPU core on Greene. 

I found that inference time of `ANN_apply_array_sio` can be further [reduced](https://github.com/Pperezhogin/MOM6/blob/ann-real-4/src/framework/MOM_ANN.F90#L465) by approximately three times:
```
"mean": 3.5530E-08 (Inference in real(4))
"mean": 3.2987E-08 (+ using Atranspose)
"mean": 3.0819E-08 (+ getting rid of normalization which is 1 for me anyway)
```
Using `real(4)` is ok as we trained ANN in `real(4)` as well. The advantage of `Atranspose` might be machine-dependent. Getting rid of normalization constant may be possible if apply normalization only with present flag (but we may skip it if it complicates the control flow too much). I might submit these modifications as we accept this PR (I might need to do some testing as 3 times faster was not expected). 

The fastest version of the code in `real(4)` precision is around _**40 Gflops**_  (`=623*2/3e-8/1e+9`, considering each ANN parameter results in two floating point operations), which is big number (40Gflops is a peak performance per core in double precision for [this](https://docs.nersc.gov/systems/perlmutter/architecture/#:~:text=39.2%20GFlops%20per%20core) processor).

File: src/framework/MOM_ANN.F90 (hunk 1 above)
Comment (adcroft) [reply src/framework/MOM_ANN.F90]: Even though all the old values of `y` are ignored, it ensures the same bit of memory is being used. We've found empirically that the compiler sometimes decides to use new memory and then copy the result which is slower. `inout` seems to be faster. 🤷

File: src/framework/MOM_ANN.F90 (hunk 2 above)
Comment (adcroft) [reply src/framework/MOM_ANN.F90]: This comment was fixed. Thx

File: src/framework/MOM_ANN.F90 (hunk 3 above)
Comment (adcroft) [reply src/framework/MOM_ANN.F90]: Good catch. Removed (force push).

File: config_src/drivers/timing_tests/time_MOM_ANN.F90 (hunk 4 above)
Comment (adcroft) [reply config_src/drivers/timing_tests/time_MOM_ANN.F90:177]: As @marshallward pointed out, what I'm calculating is the number of words for storage. When reporting Gflops, I had made the assumption that this was likely the number of multiply-adds, but as you say there is ambiguity in whether an FMA should count as one or two ops. Switching to memory processed avoided the ambiguity but in truth, it's probably better to just return the times (as we do in the other tests).
